pcdsdevices/state.py

```diff
@@ -0,0 +1,422 @@
+            event_type = self._default_sub
+        if event_type == self.SUB_STATE and not self._has_subscribed_state:
+            self.state.subscribe(self._run_sub_state, run=False)
+            self._has_subscribed_state = True
+        return cid
+
+    def _run_sub_state(self, *args, **kwargs):
+        kwargs.pop('sub_type')
+        kwargs.pop('obj')
+        self._run_subs(sub_type=self.SUB_STATE, obj=self, **kwargs)
+
+    @property
+    def position(self):
+        """
+        Name of the positioner's current state. If aliases were provided, the
+        first alias will be used instead of the base name.
+        """
+        state = self.state.get()
+        if isinstance(state, int):
+            state = self.states_enum(state).name
+        try:
+            alias = self._states_alias[state]
+            if isinstance(alias, list):
+                alias = alias[0]
+            return alias
+        except KeyError:
+            return state
+
+    @property
+    def hints(self):
+        return {'fields': [self.state.name]}
+
+    def check_value(self, value):
+        """
+        Verify that a value is a valid set state, or raise an exception.
+
+        Returns
+        -------
+        state: Enum entry
+            The corresponding Enum entry for this value. It has two meaningful
+            fields, `name` and `value`.
+        """
+        bad_value = False
+        if isinstance(value, int):
+            try:
+                state = self.states_enum(value)
+            except ValueError:
+                bad_value = True
+        elif isinstance(value, str):
+            try:
+                state = self.states_enum[value]
+            except KeyError:
+                bad_value = True
+        else:
+            raise TypeError('Valid states must be of type str or int')
+        if bad_value:
+            err = ('{0} is not a valid move state for {1}. Valid state names '
+                   'are: {2}, and their corresponding values are {3}.')
+            enum_values = [self.states_enum[state].value
+                           for state in self._valid_states]
+            raise ValueError(err.format(value, self.name, self._valid_states,
+                                        enum_values))
+        return state
+
+    def _do_move(self, state):
+        """
+        Execute the move command. Override this if your move isn't a simple put
+        to the state signal using the state name.
+
+        Parameters
+        ----------
+        state: Enum entry
+            Object whose `name` attribute is the string enum name and whose
+            `value` attribute is the integer enum value.
+        """
+        self.state.put(state.name)
+
+    def _create_states_enum(self):
+        """
+        Create an enum that can be used to keep track of aliases, state names,
+        and integer enum values.
+        """
+        state_def = {}
+        for i, state in enumerate(self.states_list):
```

Comment:
Be careful of state records that don't continuously use the `mbbi` i.e 1 is insert, 4 is removed. Maybe check if the item in the list exists so I can:
```python
_state_list = ['inserted', None, None, 'removed']
```

Reply:
Do records like these exist? I wasn't aware of any. We can definitely interpret the `None` as no state and build the `enum` accordingly. The class will also defer to your special `states_enum` if you provide one, which might be a cleaner option.